packages/sdk-middleware-http/src/http.js

```diff
@@ -64,28 +63,46 @@ export default function createHttpMiddleware({
     retryDelay = 200,
     maxDelay = Infinity,
   } = {},
+  fetch: fetcher,
 }: HttpMiddlewareOptions): Middleware {
+  if (!fetcher && typeof fetch === 'undefined')
+    throw new Error(
+      '`fetch` is not available. Please pass in `fetch` as an option or have it globally available.'
+    )
+
+  if (!fetcher)
+    // `fetcher` is set here rather than the destructuring to ensure fetch is
+    // declared before referencing it otherwise it would cause a `ReferenceError`.
+    // For reference of this pattern: https://github.com/apollographql/apollo-link/blob/498b413a5b5199b0758ce898b3bb55451f57a2fa/packages/apollo-link-http/src/httpLink.ts#L49
+
+    // eslint-disable-next-line
+    fetcher = fetch
+
   return next => (request: MiddlewareRequest, response: MiddlewareResponse) => {
     const url = host.replace(/\/$/, '') + request.uri
     const body =
       typeof request.body === 'string' || Buffer.isBuffer(request.body)
         ? request.body
-        : JSON.stringify(request.body)
+        : // NOTE: `stringify` of `null` gives the String('null')
+          JSON.stringify(request.body || undefined)
     const requestHeader = {
-      'Content-Type': 'application/json',
+      'Content-Type': ['application/json'],
       ...request.headers,
-      ...(body ? { 'Content-Length': Buffer.byteLength(body).toString() } : {}),
+      ...(body
+        ? { 'Content-Length': Buffer.byteLength(body).toString() }
+        : null),
     }
-    const requestObj: Object = new Request(url, {
+    const fetchOptions: Object = {
       method: request.method,
-      headers: new Headers(requestHeader),
+      headers: requestHeader,
       ...(credentialsMode ? { credentials: credentialsMode } : {}),
-      ...(body ? { body } : {}),
-    })
+      ...(body ? { body } : null),
+    }
     let retryCount = 0
     // wrap in a fn so we can retry if error occur
     function executeFetch() {
-      fetch(requestObj).then(
+      // $FlowFixMe
```

Comment:
I don't know this one 🤔

Reply:
It's otherwise not really type-able with Flow. At this point we know given the destructuring above and the setting of fetcher that it will always be defined or we would have thrown an error. Flow doesn't notice this. I have the same "issue" in parts with flopflip where I have to aggressively guard which we can't here. $FlowFixMe is like `eslint-ignore-next-line`.

Reply:
Thanks for the explanation